### PR TITLE
Decouple default channel insertion

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes: [#4603](https://github.com/dotnet/arcade/issues/4603)

Remove of auto insertion of default channels when the build is inserted into BAR. 

Tested locally + INT.

### CODE REVIEW

**Comment by mmitche** [comment]
@riarenas @JohnTortugo I think it's worth adding a feature flag for this (something in the settings file, defaults to off), in case we find something depending on this behavior that we don't know about.

**Comment by riarenas** [comment]
> @riarenas @JohnTortugo I think it's worth adding a feature flag for this (something in the settings file, defaults to off), in case we find something depending on this behavior that we don't know about.

Agreed. Do we have a mechanism to just upload the config files that wouldn't require a deployment? It would be great if we could just turn this on or off without triggering a full deployment of the service. 

(Or adding a more robust feature flag infra, but that is definitely waaaaay out of scope)

**Comment by mmitche** [comment]
> > @riarenas @JohnTortugo I think it's worth adding a feature flag for this (something in the settings file, defaults to off), in case we find something depending on this behavior that we don't know about.
> 
> Agreed. Do we have a mechanism to just upload the config files that wouldn't require a deployment? It would be great if we could just turn this on or off without triggering a full deployment of the service.
> 
> (Or adding a more robust feature flag infra, but that is definitely waaaaay out of scope)

I wonder whether you can do this through the UI in the Azure Portal, then restart?
Or maybe it's worth adding some REST API to set these things.

**Comment by riarenas** [comment]
> I wonder whether you can do this through the UI in the Azure Portal, then restart? 

I know you can do that for cloud services, but never tried it with service fabric. Let me spend 5 minutes looking in the portal to see if there's something obvious.

**Comment by riarenas** [comment]
Yeah, there's nothing similar to what I was hoping for in the portal.

https://github.com/Microsoft/service-fabric/issues/83 isn't very encouraging either, as even though there's such a thing as configuration packages that can be deployed independently from the code pakages, it still requires a full deployment to upgrade the configuration package.

**Comment by mmitche** [comment]
> Yeah, there's nothing similar to what I was hoping for in the portal.
> 
> [microsoft/service-fabric#83](https://github.com/microsoft/service-fabric/issues/83) isn't very encouraging either, as even though there's such a thing as configuration packages that can be deployed independently from the code pakages, it still requires a full deployment to upgrade the configuration package.

It looks like there are feature flags for aspnet. https://docs.microsoft.com/en-us/azure/azure-app-configuration/quickstart-feature-flag-aspnet-core.  It looks like that would be generally applicable.

**Comment by riarenas** [comment]
Ooh yeah, doing at the aspnet level looks promising.

**Comment by alexperovich** [comment]
We can probably use the azure App Configuration store thingy to enable feature flags, but the asp.net core integration isn't something we can use. The app isn't entirely asp.net so the feature stuff won't work outside of the website.

**Comment by mmitche** [comment]
> We can probably use the azure App Configuration store thingy to enable feature flags, but the asp.net core integration isn't something we can use. The app isn't entirely asp.net so the feature stuff won't work outside of the website.

It looks like pretty standard service configuration...maybe it's usable everywhere?  Anyways, worth a shot.

**Comment by JohnTortugo** [comment]
From the [Azure App Configuration documentation](https://docs.microsoft.com/en-us/azure/azure-app-configuration/overview):

> App Configuration makes it easier to implement the following scenarios:
> ...
> Dynamically change application settings without the need to redeploy or restart an application

**Comment by JohnTortugo** [comment]
PTAL again :-)

File: src/Maestro/Maestro.Data/Models/Build.cs [hunk 1]
```diff
@@ -22,24 +24,59 @@ public class Build
 
         static Build()
         {
+            IConfigurationRefresher configRefresher = null;
+            IConfiguration dynamicConfigs = null;
+            string appConfigurationConnString = Environment.GetEnvironmentVariable("AppConfigurationConnectionString");
```
Comment (riarenas) [inline src/Maestro/Maestro.Data/Models/Build.cs]: How is the connection string making it into the environment? Feels like this is something that should be getting pulled from a keyvault.

**Comment by riarenas** [comment]
Looks like you got the configuration store to work, nice! 

Would you mind adding some docs about which configuration stores we're going to use for each environment in case we want to add some more feature flags like this one?

File: src/Maestro/Maestro.Data/Models/Build.cs (hunk 1 above)
Comment (JohnTortugo) [reply src/Maestro/Maestro.Data/Models/Build.cs]: Initial idea was to set this on the Azure Portal but after discussion with @alexperovich and @mmitche we decided to give Azure Managed Service Identity a shot. I updated the code using that.

**Comment by JohnTortugo** [comment]
Can you all PTAL again? I updated the code to use Azure Managed Identity so we don't need to store a _connection string_, just an endpoint URI.

**Comment by riarenas** [comment]
Looks like some tests need updating.

File: src/Maestro/Maestro.Web/Startup.cs [hunk 2]
```diff
@@ -240,6 +242,28 @@ public void ConfigureServices(IServiceCollection services)
...
 
             services.AddMergePolicies();
+
+
+            // Configure access to Azure App Configuration
+            string appConfigEndpointUri = Configuration["AppConfigurationUri"];
+            ConfigurationBuilder builder = new ConfigurationBuilder();
+            TokenCredential credential = appConfigEndpointUri.Contains("maestrolocal") ? 
+                new DefaultAzureCredential() : 
+                (TokenCredential) new ManagedIdentityCredential();
```
Comment (riarenas) [inline src/Maestro/Maestro.Web/Startup.cs]: Are we sure this will use the right identity when deployed?

**Comment by riarenas** [comment]
Also added @alexperovich as a reviewer to make sure it makes sense how things are set up.

File: src/Maestro/Maestro.Web/Startup.cs (hunk 2 above)
Comment (riarenas) [reply src/Maestro/Maestro.Web/Startup.cs]: Spoke with Alex. I think you need to add the Maestro apps to the access list for the configuration store. Not dotnet-bot. Let me know if that doesn't make sense.

Comment (JohnTortugo) [reply src/Maestro/Maestro.Web/Startup.cs]: Talked with @alexperovich . I added the identities that he said should be used.

File: src/Maestro/Maestro.Data/Models/Build.cs [hunk 3]
```diff
@@ -20,26 +22,38 @@ public class Build
         private string _azureDevOpsBranch;
         private string _githubBranch;
 
+        // Used to fetch dynamic configurations from Azure App Configuration
+        // These fields are set by Startup.cs at system initialization
```
[the remark this answers is not in the document]
Comment (JohnTortugo) [reply src/Maestro/Maestro.Data/Models/Build.cs]: Done

File: src/Maestro/Maestro.Web/Startup.cs [hunk 4]
```diff
@@ -240,6 +242,28 @@ public void ConfigureServices(IServiceCollection services)
...
+
+            builder.AddAzureAppConfiguration(options =>
+            {
+                options.Connect(new Uri(appConfigEndpointUri), credential)
+                    .ConfigureRefresh(refresh =>
+                    {
+                        refresh.Register("AutoBuildPromotion", "Maestro")
+                            .SetCacheExpiration(TimeSpan.FromSeconds(1));
```
Comment (riarenas) [inline src/Maestro/Maestro.Web/Startup.cs]: Does this mean we will poll Azure for this value every second? Is such a low polling interval to the configuration store desirable?

Comment (riarenas) [reply src/Maestro/Maestro.Web/Startup.cs]: Ahh nevermind, I see now, this only happens when you call Refresh(), so this won't poll until we actually try to create a Build object.

**Comment by riarenas** [comment]
I've also been playing with the configuration Store for the feed cleanup work I'm doing, and I see that there's a separate section for app configuration and feature flags so that it's easy to turn on or off.

I think this particular one would be a good use for the feature flag feature as oposed to the text configuration that we currently have.

Take a look at https://docs.microsoft.com/en-us/azure/azure-app-configuration/manage-feature-flags it looks like the only change is to create the feature flag in the config store, and then to access it, just prefix `.appconfig.featureflag` before the flag name.

**Pull request merged**